daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -44,13 +57,112 @@ ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
                        const char **status, krb5_deltat *lifetime_out,
                        krb5_deltat *renew_lifetime_out)
 {
+    struct ipadb_context *ipactx;
+    LDAPMessage *result = NULL;
+    LDAPMessage *lentry = NULL;
+    krb5_error_code kerr;
+    char *principal = NULL;
+    enum ipadb_user_auth ua;
+    int valid_indicators_set = 0;
+
     *status = NULL;
-    *lifetime_out = 0;
-    *renew_lifetime_out = 0;
 
     krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
 
-    return 0;
+    ipactx = ipadb_get_context(context);
+    if (!ipactx) {
+        kerr = KRB5_KDB_DBNOTINITED;
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: Database not initiallized.");
+        goto done;
+    }
+
+    kerr = krb5_unparse_name(context, request->client, &principal);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: krb5_unparse_name failed.");
+        goto done;
+    }
+
+    /* Filter out anonymous PKINIT principal since it doesn't have a LDAP entry */
+    if (strncmp(principal, "WELLKNOWN/ANONYMOUS@WELLKNOWN:ANONYMOUS", 40) == 0) {
+        goto done;
+    }
+
+    /* Get LDAP entry for the principal */
+    kerr = ipadb_fetch_principals(ipactx, KRB5_KDB_FLAG_ALIAS_OK, principal, &result);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_fetch_principals failed.");
+        goto done;
+    }
+
+    kerr = ipadb_find_principal(context, KRB5_KDB_FLAG_ALIAS_OK, result, &principal, &lentry);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+        goto done;
+    }
+
+    /* Get User Auth configuration. */
+    ua = ipadb_get_user_auth(ipactx, lentry);
+
+    /* Make sure string *auth_indicators has a length of 
+     * at least 1 so that strstr() won't broke */
+    if (*auth_indicators == NULL || strlen(*auth_indicators) == 0) {
+        if (!(ua & IPADB_USER_AUTH_PASSWORD)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "Reinforced pre-authentication not allowed for this user.";
+        }
+        goto done;
+    }
+
+    /* For each auth indicator, see if it is allowed for that user */
+    if (strstr(*auth_indicators, "otp") != NULL) {
```

Comment:
This is not ok, it will match an arbitrary authenticator named "nototp" as well, you should tokenize first and then check individual tokens in their entirety.

Reply:
Just reallized auth_indicators is an array and fixed accordingly